### PR TITLE
taskmanager: remove shell interpolation from node upgrade

### LINKED ISSUE #6979: Edge node upgrade path is vulnerable to shell command injection
**What happened**:

The edge node upgrade executor builds a shell command string from unescaped request fields and runs it with `exec.Command("bash", "-c", command)`. In `edge/pkg/taskmanager/v1alpha1/taskexecutor/node_upgrade.go:194-200`, interpolated values include `upgradeReq.UpgradeID`, `upgradeReq.HistoryID`, `upgradeReq.Version`, `opts.ConfigFile`, and `upgradeReq.Image`.

**What you expected to happen**:

Upgrade execution should invoke the target command with explicit arguments and validated inputs, without routing untrusted fields through `bash -c`.

**How to reproduce it (as minimally and precisely as possible)**:

1. Open `edge/pkg/taskmanager/v1alpha1/taskexecutor/node_upgrade.go`.
2. Review lines `194-200`.
3. Confirm that user-controlled values are concatenated into a shell command string passed to `bash -c`.
4. Optionally, construct a test input containing shell metacharacters and observe that the resulting command string changes meaning.

**Anything else we need to know?**:

- Priority: P0
- Category: Security
- Why it matters: If a malicious or malformed job payload reaches this code path, shell metacharacters can execute arbitrary commands on the edge node with the privileges of the running process.
- Impact and likelihood: Impact critical, likelihood medium because exploitation depends on who can submit or modify upgrade jobs, but the sink is directly exploitable once untrusted input reaches it.
- Suggested fix: Stop using `bash -c`; invoke `exec.Command` with explicit arguments, validate allowed characters for IDs and versions, and treat image/config values as structured arguments rather than shell fragments.

**Environment**:
- Kubernetes version (use `kubectl version`): N/A
- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`): Affected current repository state; exact release not yet bounded
- <details><summary>Cloud nodes Environment:</summary>

  - Hardware configuration (e.g. `lscpu`): N/A
  - OS (e.g. `cat /etc/os-release`): N/A
  - Kernel (e.g. `uname -a`): N/A
  - Go version (e.g. `go version`): N/A
  - Others: Issue is triggered by upgrade job construction, not cloud host specifics.

  </details>
- <details><summary>Edge nodes Environment:</summary>

  - edgecore version (e.g. `edgecore --version`): Affected current repository state; exact release not yet bounded
  - Hardware configuration (e.g. `lscpu`): N/A
  - OS (e.g. `cat /etc/os-release`): Any OS that runs this upgrade path
  - Kernel (e.g. `uname -a`): N/A
  - Go version (e.g. `go version`): N/A
  - Others: Exploitability depends on whether an attacker can influence upgrade job fields.

  </details>

### PR DESCRIPTION
**Pull request by rootp1**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Removes the untrusted `bash -c` interpolation from the v1alpha1 edge node upgrade executor and replaces the non-systemd path with an explicit `keadm` argv invocation. On systemd-managed nodes, the upgrade is launched through a transient `systemd-run` unit and a fixed shell wrapper that reconstructs the final `keadm` argv from environment variables, so request fields are not interpolated into shell source and the upgrade survives the EdgeCore service stop during self-upgrade. The patch also hardens the non-systemd log-file path under `/var/log/kubeedge`, restores truncating log behavior for each upgrade attempt, and uses journald as the default logging destination for the transient systemd unit.

**Which issue(s) this PR fixes**:
Fixes #6979

**Special notes for your reviewer**:
Added focused unit coverage for argv construction, transient systemd launch construction, secure log creation, symlink rejection, log truncation behavior, transient unit identity uniqueness, and conditional `systemd-run` capability handling.

Verified locally following `CONTRIBUTING.md` with:
- `GOCACHE=/Users/rootp1/Documents/repos/kubeedge/.worktrees/pr6980-vm/.cache/go-build go test ./edge/pkg/taskmanager/v1alpha1/taskexecutor`
- `GOCACHE=/Users/rootp1/Documents/repos/kubeedge/.worktrees/pr6980-vm/.cache/go-build go test -race ./edge/pkg/taskmanager/v1alpha1/taskexecutor`
- `GOOS=windows GOARCH=amd64 go build ./edge/cmd/edgecore`
- `PATH="/opt/homebrew/opt/md5sha1sum/bin:/opt/homebrew/bin:$PATH" make verify`

Verified the systemd lifecycle in a real Linux systemd VM:
- Ubuntu 24.04.4 LTS guest
- repository `build/tools/edgecore.service`
- upgrade launched through the v1alpha1 `keadmUpgrade(...)` path
- confirmed the upgrade process survived `systemctl stop edgecore.service`
- confirmed binary replacement completed
- confirmed `edgecore.service` restarted successfully on the replaced binary

Verified additional real-systemd regressions in the same VM:
- exact hostile argv values reached `keadm` literally through the real systemd manager
- a failed transient unit did not block a retry with the same `UpgradeID` and a new `HistoryID`

Evidence from the lifecycle run:
- `systemd-run` launched the transient unit and the helper `keadm` process survived `systemctl stop edgecore.service`
- `/var/log/kubeedge/lifecycle.log` recorded `before-stop`, `survived-stop`, and `restart-complete`
- `/var/log/kubeedge/edgecore-lifecycle.log` recorded `edgecore-v1-start`, `edgecore-v1-stop`, and `edgecore-v2-start`
- final `systemctl status edgecore.service` showed the restarted service active on the replaced binary

**Does this PR introduce a user-facing change?**:
```release-note
Fixed a command injection risk in the v1alpha1 edge node upgrade flow by removing untrusted shell interpolation from the upgrade launch path. Non-systemd nodes now invoke `keadm` with explicit argv values, while systemd-managed nodes launch the upgrade in a transient systemd unit using a fixed shell wrapper with environment-provided arguments. Non-systemd upgrade logs are written safely under `/var/log/kubeedge`.
```

### CODE REVIEW

File: edge/pkg/taskmanager/v1alpha1/taskexecutor/node_upgrade.go [hunk 1]
```diff
@@ -191,21 +193,52 @@ func upgrade(taskReq types.NodeTaskRequest) (event fsm.Event) {
...
+		if cerr := logFile.Close(); cerr != nil {
+			klog.Warningf("failed to close upgrade log file %s: %v", keadmUpgradeLog, cerr)
+		}
+	}()
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("start upgrade command %v failed: %w", cmd.Args, err)
+	}
+	if err := cmd.Process.Release(); err != nil {
+		klog.Warningf("failed to release upgrade command process %v: %v", cmd.Args, err)
 	}
```
[the remark this answers is not in the document]
Comment (rootp1) [reply edge/pkg/taskmanager/v1alpha1/taskexecutor/node_upgrade.go]: Applied. I replaced `Process.Release()` with a background `cmd.Wait()` helper so the detached upgrade process is still reaped when it exits.

**Comment by DoisLONG** [comment]
Thanks for replacing the shell command with a direct `keadm` invocation. The explicit argv construction removes the command-injection sink, the absolute executable path avoids PATH-based command selection, and the background `cmd.Wait()` correctly reaps the child when EdgeCore remains alive.

After following the full upgrade lifecycle, I found two important issues that still need to be addressed.

### 1. Verify that the upgrade process survives stopping EdgeCore under systemd

The new command is still launched as a child of the EdgeCore process:

```go
cmd := exec.Command(keadmPath, args...)
cmd.Start()
go waitForUpgradeCommand(cmd, logFile)
```

During the upgrade, `keadm upgrade edge` calls `KillKubeEdgeBinary`, which uses `systemctl stop <edgecore-service>` when EdgeCore is managed by systemd.

The repository's `edgecore.service` does not override `KillMode`, so the systemd default `KillMode=control-group` applies. Stopping the service terminates the remaining processes in the unit's control group, which can include the child `keadm` process itself.

`nohup` did not create a separate systemd cgroup either, so this may be a pre-existing lifecycle defect, but the current unit tests do not demonstrate that the claimed detached upgrade behavior is preserved.

Please validate the complete flow using the standard systemd-managed EdgeCore installation:

1. start EdgeCore through the repository service unit;
2. trigger the v1alpha1 upgrade;
3. verify that the upgrade process survives the EdgeCore service stop;
4. verify that the binary replacement and EdgeCore restart complete successfully.

If the child is terminated with the service, it needs to be launched through a genuinely independent supervisor or transient systemd unit rather than only through a goroutine.

### 2. Do not open a privileged fixed log file in `/tmp`

The current code opens:

```go
os.OpenFile(
    "/tmp/keadm.log",
    os.O_CREATE|os.O_WRONLY|os.O_APPEND,
    0o600,
)
```

A local user can pre-create `/tmp/keadm.log` as a symbolic link. Since the open operation does not reject symlinks, a privileged EdgeCore process may append upgrade output to an attacker-selected target.

Also, `0o600` applies only when the file is newly created. It does not correct the permissions or ownership of an existing file.

Please place the log under a protected KubeEdge-owned directory such as `/var/log/kubeedge`, reject non-regular files or symlinks, and ensure the final ownership and permissions are correct.

### 3. Preserve or explicitly redefine the log behavior

The previous shell redirection used `>`, which truncated the log for each upgrade. The new `O_APPEND` behavior accumulates all upgrade output indefinitely and may mix multiple upgrade attempts.

Please either:

* preserve the previous behavior with a safely opened truncated file; or
* use a protected per-upgrade log file with an explicit retention strategy.

Please add regression coverage for the secure log-file behavior and provide the systemd upgrade validation result. Since this is a public security fix, the release note should also describe the command-injection remediation rather than using `NONE`.

**Comment by rootp1** [comment]
Addressed in the latest update.

1. The systemd-managed path now launches the upgrade through a transient `systemd-run` unit instead of leaving `keadm` as a child of EdgeCore. That avoids tying the upgrade process to the EdgeCore service cgroup during `systemctl stop`.

2. The upgrade log is no longer opened in `/tmp`. It now uses `/var/log/kubeedge/keadm-upgrade.log`, creates the parent directory with protected permissions, rejects symlink and non-regular log targets, and enforces `0600` on the file.

3. The log behavior is truncating again for each upgrade attempt, which matches the previous `>` semantics.

I also added regression coverage for:
- direct argv construction
- transient systemd launch construction
- symlink rejection
- protected log creation
- truncation of existing log content

Validated locally with:
- `GOCACHE=/Users/rootp1/Documents/repos/kubeedge/.cache/go-build go test ./edge/pkg/taskmanager/v1alpha1/taskexecutor`
- `PATH="/opt/homebrew/opt/md5sha1sum/bin:/opt/homebrew/bin:$PATH" make verify`

**Comment by rootp1** [comment]
Systemd lifecycle validation is now completed in a real Linux systemd VM.

Environment used:
- Ubuntu 24.04.4 LTS guest started with Lima on this host
- repository `build/tools/edgecore.service`
- systemd running inside the guest

Validation flow executed:
1. Installed the repository `edgecore.service` into `/etc/systemd/system/edgecore.service` and started it with `systemctl restart edgecore.service`
2. Triggered the v1alpha1 launcher path by running the package test harness that calls `keadmUpgrade(...)` inside the guest
3. The Linux path launched `keadm` through `systemd-run` as expected
4. The fake upgrade command stopped `edgecore.service`, survived the stop, replaced `/usr/local/bin/edgecore`, and restarted the service successfully

Observed result:
- `keadmUpgrade` launched: `/usr/bin/systemd-run --unit kubeedge-keadm-upgrade-lifecycle-upgrade-1 ... /usr/local/bin/keadm upgrade edge ...`
- `/var/log/kubeedge/lifecycle.log` contained:
  - `before-stop ...`
  - `survived-stop ...`
  - `restart-complete ...`
- `/var/log/kubeedge/edgecore-lifecycle.log` contained:
  - `edgecore-v1-start ...`
  - `edgecore-v1-stop ...`
  - `edgecore-v2-start ...`
- final `systemctl status edgecore.service` showed the restarted service active on the replaced binary

The in-guest lifecycle test passed:
`go test -run TestKeadmUpgradeLifecycleUnderSystemd -v ./edge/pkg/taskmanager/v1alpha1/taskexecutor`

This validates that the current Linux systemd path survives the EdgeCore service stop because the upgrade runs in a transient systemd unit rather than as a child in the EdgeCore service cgroup.

**Comment by rootp1** [comment]
@DoisLONG please have a look whenever you have a chance to

**Comment by DoisLONG** [comment]
Thanks for the substantial update. The previous review items have been addressed well:

* the shell invocation has been removed;
* the non-systemd child is reaped with `Wait`;
* systemd-managed upgrades run in a transient unit and have been validated in a real VM;
* the log was moved out of `/tmp`;
* truncating behavior and the release note have been restored.

I found several remaining issues in the new transient-unit path.

### 1. `systemd-run` does not preserve the arguments literally by default

Although the command no longer passes through a shell, `systemd-run` enables environment-variable expansion in command arguments by default. In service mode, this expansion is performed by the service manager.

As a result, values such as:

```text
example.com/package:${PATH}
${UNSET_VARIABLE}
```

may be expanded, removed, or split before they reach `keadm`. This violates the security requirement that every request field be passed as one unchanged argv value.

The current metacharacter test covers shell constructs such as `;`, `&&`, and `$(...)`, but it does not cover systemd's `${VARIABLE}` expansion.

Please add an integration-style transient-unit test that places `${KUBEEDGE_TEST_VALUE}` in every externally influenced argument and verifies the exact argv received by the launched helper.

`--expand-environment=no` would disable this behavior, but that option was added only in systemd v254, so simply adding it would introduce a very high minimum-version requirement. The implementation needs a literal-argv design that works across the systemd versions KubeEdge supports.

### 2. Detect the required systemd capabilities, not only the binary

`canUseSystemdRun` currently checks only that systemd is running and that `systemd-run` exists.

However, the invocation depends on newer functionality. For example, `--collect` was introduced in systemd v236. A host may have a working `systemd-run` binary but not support one or more of the options or output properties used here.

Please either:

* use a command compatible with the minimum supported systemd version;
* detect the required capabilities/version before selecting this path;
* or define and validate a supported fallback that still survives stopping `edgecore.service`.

Falling back to a normal child process on a systemd-managed node is not sufficient, because that child may remain in the EdgeCore unit's control group.

### 3. Make the log-file protection resistant to replacement races

The current protection performs `Lstat` and `OpenFile` as separate path operations. In the systemd path, `prepareUpgradeLogFile` then closes the file and systemd opens the pathname again through `StandardOutput=append:`.

The path can therefore change after validation.

Also, `ensureSecureLogDir` changes only the mode and does not verify or correct the directory owner. `openUpgradeLogFile` similarly changes only the file mode. If an existing `/var/log/kubeedge` directory is owned by an untrusted local account, that owner can still replace the log entry after the checks.

Please ensure that:

* the log directory has the expected privileged owner and is not writable by untrusted users;
* the final log file has the expected owner and mode;
* the direct-open path uses atomic no-follow semantics;
* the path reopened by systemd cannot be replaced after validation.

Using journald for the transient unit would avoid the second path-open operation entirely.

### 4. Bound and deconflict the transient unit name

`UpgradeID` is derived from `NodeUpgradeJob.Name`, and the code adds a fixed prefix before systemd appends `.service`.

Please ensure the final unit name always fits systemd's unit-name limit. A stable form such as:

```text
readable-prefix + truncated-ID-prefix + hash(original-ID)
```

would preserve readability while preventing truncation collisions.

Please add tests for empty, long, and same-prefix/different-ID values.

### 5. Add the repository license header

The new `node_upgrade_test.go` file is missing the standard KubeEdge Apache 2.0 boilerplate.

After these points are addressed, please rerun the focused tests, `make verify`, and the real systemd lifecycle test.

**Comment by rootp1** [comment]
Thanks for the detailed follow-up. I addressed the remaining points in the latest update.

1. The transient-unit path no longer passes request-derived values to `systemd-run` as command arguments. Instead, the unit receives them through `--setenv`, and a fixed `/bin/sh -c` wrapper reconstructs the final `keadm` argv from those environment variables. That keeps values such as `${KUBEEDGE_TEST_VALUE}` literal all the way through to `keadm`.

2. I tightened systemd capability selection. The Linux systemd path now verifies that `systemd-run` supports the feature it relies on (`--setenv`) before using the transient-unit launch path, rather than checking only for the binary.

3. I reworked the non-systemd log-file handling to use atomic no-follow open semantics and owner checks:
   - the log directory must not be a symlink and must be owned by the current privileged user
   - the log file is opened with `O_NOFOLLOW`
   - the opened handle is verified to be a regular file owned by the current user
   - permissions are enforced to `0750` for the directory and `0600` for the file

   For the transient systemd path, output no longer depends on reopening a validated pathname through `StandardOutput=append:`.

4. The transient unit name is now bounded and deconflicted using a stable hashed suffix, with tests covering empty IDs, long IDs, and same-prefix/different-ID cases.

5. I added the standard Apache 2.0 header to `node_upgrade_test.go`.

Validation rerun:
- `GOCACHE=/Users/rootp1/Documents/repos/kubeedge/.worktrees/pr6980-vm/.cache/go-build go test ./edge/pkg/taskmanager/v1alpha1/taskexecutor`
- `PATH="/opt/homebrew/opt/md5sha1sum/bin:/opt/homebrew/bin:$PATH" make verify`

Real Linux systemd lifecycle validation rerun:
- Ubuntu 24.04.4 LTS guest in Lima
- repository `build/tools/edgecore.service`
- `systemd 255`
- privileged package-local harness invoking `keadmUpgrade(...)` on the current branch head

Observed result from the rerun:
- transient unit launched as `kubeedge-keadm-upgrade-lifecycle-upgrade-1-4538791bf59c.service`
- `/var/log/kubeedge/keadm-upgrade.log` recorded:
  - `fake-keadm start ... args:upgrade edge --upgradeID lifecycle-upgrade-1 --historyID history-1 --fromVersion ... --toVersion v1.17.0 --config /etc/kubeedge/edgecore.yaml --image example.com/installation-package:v1.17.0`
- `/var/log/kubeedge/lifecycle.log` recorded:
  - `before-stop ...`
  - `survived-stop ...`
  - `restart-complete ...`
- `/var/log/kubeedge/edgecore-lifecycle.log` recorded:
  - `edgecore-v1-stop ...`
  - `edgecore-v1-start ...`
  - `edgecore-v1-stop ...`
  - `edgecore-v2-start ...`
- final `systemctl status edgecore.service` showed the restarted service active on the replaced binary

That rerun confirms the current Linux systemd path survives the `edgecore.service` stop, completes the replacement, and restarts the service successfully with the updated transient-unit launch design.

**Comment by DoisLONG** [comment]
Thanks for the latest update. The new revision addresses several earlier concerns well, including process reaping, unit-name length and collision handling, `O_NOFOLLOW`, ownership checks, the test-file license header, and a literal-argument test.

I found several remaining issues.

### 1. Preserve Windows build compatibility

`node_upgrade.go` is a platform-independent file, but it now imports `golang.org/x/sys/unix` and uses `unix.Open`, Unix flags, and `syscall.Stat_t`.

The runtime `GOOS` check does not help because the file must compile before that branch can execute. This will break Windows EdgeCore builds.

Please move the platform-specific file-opening and ownership logic into build-tagged files, for example:

* `node_upgrade_log_linux.go`;
* `node_upgrade_log_windows.go`.

Please verify the result with a Windows cross-build such as:

`GOOS=windows GOARCH=amd64 go build ./edge/cmd/edgecore`

### 2. Allow retry after a failed transient unit

The transient unit name is derived only from `UpgradeID`, and the current command no longer uses `--collect`.

Without collection, systemd keeps failed transient units loaded until their failure state is reset. Retrying or recreating a NodeUpgradeJob with the same name can therefore attempt to create the same unit and fail because it already exists.

`HistoryID` is generated uniquely for each legacy upgrade dispatch, so please include it in the unit-name identity, conditionally use `--collect` where supported, or explicitly clean up the failed unit.

Please add a real-systemd regression test where the first invocation fails and a second invocation with the same UpgradeID can still start.

### 3. Make the logging behavior consistent with the PR description

The non-systemd path writes to `/var/log/kubeedge/keadm-upgrade.log`, but the systemd path does not configure file output and will log through journald.

Please either:

* document journald as the systemd logging destination and update the PR body/release note;
* or implement a secure file-output mechanism for the transient unit.

The current statement that upgrade logs are written under `/var/log/kubeedge` is not true for the default systemd path.

### 4. Validate literal argv through the real systemd manager

The current test uses a shell script that simulates `systemd-run`. It does not exercise systemd's own command-argument environment expansion.

Because the security guarantee depends on the interaction between systemd expansion and `/bin/sh -c`, please add a real-systemd exact-argv test with values containing spaces, quotes, semicolons, command substitutions, `$VAR`, `${VAR}`, backslashes, and newlines.

The test should verify the exact argument count and byte content and confirm that no command side effects occur.

### 5. Validate the log target before truncating it

`openUpgradeLogFile` currently opens the target with `O_TRUNC` before checking its type and owner.

A target with an unexpected owner is therefore truncated before the function rejects it.

Please open without `O_TRUNC`, validate the opened file with `fstat`, and only then truncate it. Add a test confirming that rejected files retain their original content.

### 6. Update the title or implementation description

The systemd path still invokes `/bin/sh -c`. The security improvement is that untrusted values are no longer interpolated into the shell source, not that every shell invocation has been removed.

Please update the title and documentation to describe that distinction accurately.

After these changes, please provide:

* focused package tests;
* race-enabled tests;
* a Windows cross-build;
* a real-systemd failure-and-retry test;
* a real-systemd malicious-argument exact-argv test.

**Comment by rootp1** [comment]
Thanks. I addressed the remaining issues in the latest update.

1. Windows build compatibility is restored. The Unix-specific log opening and ownership checks now live in build-tagged files, and `node_upgrade.go` is platform-independent again.

2. The transient-unit identity now includes both `UpgradeID` and `HistoryID`, and the systemd path uses `--collect` when the host `systemd-run` supports it. That keeps failed transient units from blocking a later retry, and the distinct `HistoryID` also prevents name reuse across legacy redispatches.

3. I updated the PR title/body and release note text to match the implementation precisely:
   - non-systemd path: direct `keadm` argv + secure `/var/log/kubeedge/keadm-upgrade.log`
   - systemd path: transient unit + fixed shell wrapper + journald as the default log destination

4. The real-systemd exact-argv validation was rerun under the actual systemd manager with hostile values containing spaces, quotes, semicolons, command substitutions, `$VAR`, `${VAR}`, backslashes, and newlines. The helper `keadm` process received the exact literal argv values, and the sentinel command-substitution side effects (`/tmp/pwned*`) were not created.

5. `openUpgradeLogFile` no longer truncates before validation. It now:
   - opens without `O_TRUNC`
   - validates the opened file with `fstat`/ownership checks
   - only then truncates and rewinds the accepted file

   I added coverage to confirm rejected targets retain their original content.

6. I updated the wording to describe the security improvement accurately: the systemd path still uses `/bin/sh -c`, but untrusted request values are no longer interpolated into shell source.

Validation rerun:
- `go test ./edge/pkg/taskmanager/v1alpha1/taskexecutor`
- `go test -race ./edge/pkg/taskmanager/v1alpha1/taskexecutor`
- `GOOS=windows GOARCH=amd64 go build ./edge/cmd/edgecore`
- `make verify`

Real-systemd regression reruns in the Ubuntu 24.04.4 Lima VM:
- exact hostile argv test: passed through the real manager with literal byte preservation and no command side effects
- failure/retry test: first transient unit failed, second launch with the same `UpgradeID` and a new `HistoryID` started successfully
- lifecycle test: rerun passed with transient unit `kubeedge-keadm-upgrade-lifecycle-upgrade-1-history-1-4188c9c59535`, `/var/log/kubeedge/lifecycle.log` recorded `before-stop`, `survived-stop`, and `restart-complete`, and the restarted `edgecore.service` came back active on the replaced binary

**Comment by rootp1** [comment]
@DoisLONG please have a look whenever you have a chance to

**Comment by rootp1** [comment]
@DoisLONG following up on this one, the Windows build fix, retry after failed transient unit, logging consistency, literal argv validation under real systemd, and the truncate before validate ordering issue should all be addressed as described in my last update. Let me know if you want me to revalidate anything.

**Comment by rootp1** [comment]
@DoisLONG merged master in and resolved the conflicts. Master had independently added a simpler direct exec.Command version of the same v1alpha1 launcher without the systemd transient unit handling, so I kept this branch's fuller launcher implementation, systemd path, log protections, and unit naming, and just reconciled the surrounding import and function boundary changes. Ran the taskexecutor package tests plus a Windows cross build of edgecore and both passed.

**Comment by rootp1** [comment]
@DoisLONG the Windows build tag split, retry-after-failure unit naming, log truncate-before-validate ordering, and the real-systemd argv checks from your last review are all in place. I squashed the branch to a single commit and merged the current master in to clear a conflict that came up in node_upgrade.go, keeping this branch's systemd transient-unit path. Reran the taskexecutor package tests and a Windows cross build (GOOS=windows GOARCH=amd64), both pass. Ready for a re-review whenever you get a chance.

**Comment by rootp1** [comment]
@DoisLONG this one has survived multiple rounds of thorough review and all the validation, including the real systemd testing, has passed. I would be grateful if this could be merged soon.

**Comment by rootp1** [comment]
@DoisLONG thanks. I squashed this branch to a single commit now. The code itself is unchanged from the version you already reviewed. Please take another look whenever you have time.

**Comment by DoisLONG** [comment]
Thanks for the continued updates.

CI currently looks like it needs to be triggered/approved, but I do not see any remaining code blocker from my side.

/lgtm

**Comment by rootp1** [comment]
@DoisLONG lgtm got cleared after I amended the commit message to a subject-only form for convention, the diff you reviewed did not change. Please have another look and re-add lgtm when you can.

**Review by DoisLONG — COMMENTED** [review]
/lgtm
